runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClient.java

```diff
@@ -231,122 +234,114 @@ public ActiveBundle newBundle(
+      private final StateDelegator.Registration stateRegistration;
+      private final BundleProgressHandler progressHandler;
+      private final BundleSplitHandler splitHandler;
+      private final BundleCheckpointHandler checkpointHandler;
+      private final BundleFinalizationHandler finalizationHandler;
+
+      private ActiveBundle(
+          String bundleId,
+          CompletionStage<ProcessBundleResponse> response,
+          Map<String, CloseableFnDataReceiver> inputReceivers,
+          Map<String, InboundDataClient> outputClients,
+          StateDelegator.Registration stateRegistration,
+          BundleProgressHandler progressHandler,
+          BundleSplitHandler splitHandler,
+          BundleCheckpointHandler checkpointHandler,
+          BundleFinalizationHandler finalizationHandler) {
+        this.bundleId = bundleId;
+        this.response = response;
+        this.inputReceivers = inputReceivers;
+        this.outputClients = outputClients;
+        this.stateRegistration = stateRegistration;
+        this.progressHandler = progressHandler;
+        this.splitHandler = splitHandler;
+        this.checkpointHandler = checkpointHandler;
+        this.finalizationHandler = finalizationHandler;
+      }
 
-    /** Returns an id used to represent this bundle. */
-    @Override
-    public String getId() {
-      return bundleId;
-    }
+      /** Returns an id used to represent this bundle. */
+      @Override
+      public String getId() {
+        return bundleId;
+      }
 
-    /**
-     * Get a map of PCollection ids to {@link FnDataReceiver receiver}s which consume input
-     * elements, forwarding them to the remote environment.
-     */
-    @Override
-    public Map<String, FnDataReceiver> getInputReceivers() {
-      return (Map) inputReceivers;
-    }
+      /**
+       * Get a map of PCollection ids to {@link FnDataReceiver receiver}s which consume input
+       * elements, forwarding them to the remote environment.
+       */
+      @Override
+      public Map<String, FnDataReceiver> getInputReceivers() {
+        return (Map) inputReceivers;
+      }
 
-    /**
-     * Blocks until bundle processing is finished. This is comprised of:
-     *
-     * <ul>
-     *   <li>closing each {@link #getInputReceivers() input receiver}.
-     *   <li>waiting for the SDK to say that processing the bundle is finished.
-     *   <li>waiting for all inbound data clients to complete
-     * </ul>
-     *
-     * <p>This method will throw an exception if bundle processing has failed. {@link
-     * Throwable#getSuppressed()} will return all the reasons as to why processing has failed.
-     */
-    @Override
-    public void close() throws Exception {
-      Exception exception = null;
-      for (CloseableFnDataReceiver<?> inputReceiver : inputReceivers.values()) {
+      @Override
+      public void split(double fractionOfRemainder) {
+        throw new UnsupportedOperationException("Support splitting, TODO: BEAM-5600");
```

Comment:
I need access to the control service which is why I made the ActiveBundle a nested class.